src/shell/shell_cmds.cpp

```diff
@@ -578,86 +611,114 @@ void DOS_Shell::CMD_DIR(char * args) {
+	for (auto &entry : results) {
 
-		char * name = iter->name;
-		Bit32u size = iter->size;
-		Bit16u date = iter->date;
-		Bit16u time = iter->time;
-		Bit8u attr = iter->attr;
+		char *name = entry.name;
+		const uint32_t size = entry.size;
+		const uint16_t date = entry.date;
+		const uint16_t time = entry.time;
+		const bool is_dir = entry.attr & DOS_ATTR_DIRECTORY;
 
-		/* output the file */
-		if (optB) {
-			// this overrides pretty much everything
-			if (strcmp(".",name) && strcmp("..",name)) {
-				WriteOut("%s\n",name);
-			} else {
-				// skip to the next file, otherwise this file
-				// will be counted as printed for pause cmd
+		// Skip listing . and .. from toplevel directory, to simulate
+		// DIR output correctly.
+		// Bare format never lists .. nor . as directories.
+		if (is_root || optB) {
+			if (strcmp(".", name) == 0 || strcmp("..", name) == 0)
 				continue;
-			}
+		}
+
+		if (is_dir) {
+			dir_count += 1;
 		} else {
-			char * ext = empty_string;
-			if (!optW && (name[0] != '.')) {
-				ext = strrchr(name, '.');
-				if (!ext) ext = empty_string;
-				else *ext++ = 0;
-			}
-			Bit8u day	= (Bit8u)(date & 0x001f);
-			Bit8u month	= (Bit8u)((date >> 5) & 0x000f);
-			Bit16u year = (Bit16u)((date >> 9) + 1980);
-			Bit8u hour	= (Bit8u)((time >> 5 ) >> 6);
-			Bit8u minute = (Bit8u)((time >> 5) & 0x003f);
-
-			if (attr & DOS_ATTR_DIRECTORY) {
-				if (optW) {
-					WriteOut("[%s]",name);
-					size_t namelen = strlen(name);
-					if (namelen <= 14) {
-						for (size_t i=14-namelen;i>0;i--) WriteOut(" ");
-					}
-				} else {
-					WriteOut("%-8s %-3s   %-16s %02d-%02d-%04d %2d:%02d\n",name,ext,"<DIR>",day,month,year,hour,minute);
-				}
-				dir_count++;
+			file_count += 1;
+			byte_count += size;
+		}
+
+		// 'Bare' format: just the name, one per line, nothing else
+		//
+		if (optB) {
+			WriteOut("%s\n", name);
+			show_press_any_key();
+			continue;
+		}
+
+		// 'Wide list' format: using several columns
+		//
+		if (optW) {
+			if (is_dir) {
+				const size_t namelen = strlen(name);
+				WriteOut("[%s]%*s", name, (14 - namelen), "");
 			} else {
-				if (optW) {
-					WriteOut("%-16s",name);
-				} else {
-					FormatNumber(size,numformat);
-					WriteOut("%-8s %-3s   %16s %02d-%02d-%04d %2d:%02d\n",name,ext,numformat,day,month,year,hour,minute);
-				}
-				file_count++;
-				byte_count+=size;
+				WriteOut("%-16s", name);
 			}
-			if (optW) {
-				w_count++;
+			w_count += 1;
+			if ((w_count % 5) == 0)
+				show_press_any_key();
+			continue;
+		}
+
+		// default format: one detailed entry per line
+		//
+		const auto year   = static_cast<uint16_t>((date >> 9) + 1980);
+		const auto month  = static_cast<uint8_t>((date >> 5) & 0x000f);
+		const auto day    = static_cast<uint8_t>(date & 0x001f);
+		const auto hour   = static_cast<uint8_t>((time >> 5) >> 6);
+		const auto minute = static_cast<uint8_t>((time >> 5) & 0x003f);
```

Comment:
File dates on pre-1980s games (_Mission Impossible_ and [_Dunjonquest: Morloc's Tower_](https://en.wikipedia.org/wiki/List_of_PC_booter_games)) are wrapped to the year 2100+. 

Given the issue exists in SVN, it's you're call if you want to ignore this.  Also, because DOS [emerged in 1980](https://en.wikipedia.org/wiki/Timeline_of_DOS_operating_systems) the only pre-1980s games would have been PC-Booters; so the only way to list their content would be have a Booter-to-DOS conversion that retained some of the original file dates.

That said, if you bought a PC in 1980, I would think it absurd if the OS was incapable of listing dates properly for the prior year.  Surely DOS itself took a year or more to develop, so developers would have been testing it on pre-1980s files.  So my guess is original DOS would have handled this correctly.

![2019-12-26_09-33](https://user-images.githubusercontent.com/1557255/71484858-62ec9800-27c3-11ea-8b01-cd48eb7342bf.png)

Reply:
*The MS-DOS date format can represent only dates between 1/1/1980 and 12/31/2107* ([MSDN](https://docs.microsoft.com/en-gb/windows/win32/api/winbase/nf-winbase-filetimetodosdatetime?redirectedfrom=MSDN)).

I guess backwards compatibility was not an important issue back then.

Reply:
Thanks for the  reference:  "today I learned", as they say.  Hilarious they just ruled out handling dates correctly from even the prior year! As if some company wouldn't want to store and date recorda from prior years :-).  Those were the good ole' days.

Reply:
At least they won't have to deal with Y2038-like problem for several more decades ;)